# 빅데이터 분석 특강 (11주차) 5월16일

- toc:true
- branch: master
- badges: true
- comments: false
- author: 최서연
- categories: [Special Topics in Big Data Analysis]

### imports

In [1]:
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np
import tensorflow.experimental.numpy as tnp

In [2]:
tf.config.experimental.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [3]:
tnp.experimental_enable_numpy_behavior()

In [4]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

In [5]:
y_train # 이 데이터는 클래스의 데이터

array([9, 0, 0, ..., 3, 0, 5], dtype=uint8)

In [6]:
x_train.shape

(60000, 28, 28)

In [10]:
X = tf.constant(x_train.reshape(-1,28,28,1),dtype=tf.float64)
y = tf.keras.utils.to_categorical(y_train)
XX = tf.constant(x_test.reshape(-1,28,28,1),dtype=tf.float64)
yy = tf.keras.utils.to_categorical(y_test)

In [15]:
net1 = tf.keras.Sequential()
net1.add(tf.keras.layers.Flatten())
net1.add(tf.keras.layers.Dense(30,activation='relu')) # 30 잘 안 맞는 거 아는데 처음이니까~ 해보자
net1.add(tf.keras.layers.Dense(30,activation='relu'))
net1.add(tf.keras.layers.Dense(10,activation='softmax'))
net1.compile(optimizer='adam',loss=tf.losses.categorical_crossentropy,metrics='accuracy')
net1.fit(X,y,epochs=10)

Epoch 1/10
  57/1875 [..............................] - ETA: 5s - loss: 17.1517 - accuracy: 0.2111  

2022-05-16 19:18:57.584523: I tensorflow/stream_executor/cuda/cuda_blas.cc:1774] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.


1875/1875 [==============================] - 7s 3ms/step - loss: 1.9477 - accuracy: 0.4426
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 1.0451 - accuracy: 0.5917
Epoch 3/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.8487 - accuracy: 0.6686
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.7931 - accuracy: 0.6836
Epoch 5/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.7587 - accuracy: 0.7028
Epoch 6/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.7128 - accuracy: 0.7225
Epoch 7/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6896 - accuracy: 0.7284
Epoch 8/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6661 - accuracy: 0.7359
Epoch 9/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6561 - accuracy: 0.7393
Epoch 10/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6475 - accurac

In [16]:
net1.evaluate(XX,yy)

313/313 [==============================] - 1s 3ms/step - loss: 0.7311 - accuracy: 0.7128


[0.7310638427734375, 0.7128000259399414]

net가 단순하다보니 정확도 계산에 한계가 있는 거 같다 에폭 20해도 크게 변하지 않음!

In [44]:
net1 = tf.keras.Sequential()
net1.add(tf.keras.layers.Flatten())
net1.add(tf.keras.layers.Dense(500,activation='relu'))
net1.add(tf.keras.layers.Dense(500,activation='relu'))
net1.add(tf.keras.layers.Dense(10,activation='softmax'))
net1.compile(optimizer='adam',loss=tf.losses.categorical_crossentropy,metrics='accuracy')
net1.fit(X,y,epochs=10)

Epoch 1/10
1875/1875 [==============================] - 7s 4ms/step - loss: 2.0995 - accuracy: 0.7383
Epoch 2/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.7340 - accuracy: 0.7693
Epoch 3/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.5704 - accuracy: 0.8125
Epoch 4/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4953 - accuracy: 0.8313
Epoch 5/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4669 - accuracy: 0.8410
Epoch 6/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4544 - accuracy: 0.8445
Epoch 7/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4325 - accuracy: 0.8504
Epoch 8/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4320 - accuracy: 0.8501
Epoch 9/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4231 - accuracy: 0.8539
Epoch 10/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.418

In [45]:
net1.evaluate(XX,yy)

313/313 [==============================] - 1s 3ms/step - loss: 0.4547 - accuracy: 0.8397


[0.4547102153301239, 0.8396999835968018]

In [46]:
net1.summary()

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_8 (Flatten)         (None, 784)               0         
                                                                 
 dense_20 (Dense)            (None, 500)               392500    
                                                                 
 dense_21 (Dense)            (None, 500)               250500    
                                                                 
 dense_22 (Dense)            (None, 10)                5010      
                                                                 
Total params: 648,010
Trainable params: 648,010
Non-trainable params: 0
_________________________________________________________________


Dense layer로 한계가 있는 것 같으니 다른 방법을 찾아보자

---

`-` 두번째 시도
- `tf.keras.layers.Conv2D`
- `tf.keras.layers.MaxPool2D`

In [47]:
net2 = tf.keras.Sequential()
net2.add(tf.keras.layers.Conv2D(30,(2,2)))
net2.add(tf.keras.layers.MaxPool2D())
net2.add(tf.keras.layers.Conv2D(30,(2,2)))
net2.add(tf.keras.layers.MaxPool2D())
net2.add(tf.keras.layers.Flatten())
net2.add(tf.keras.layers.Dense(10,activation='softmax'))
net2.compile(optimizer='adam',loss=tf.losses.categorical_crossentropy,metrics='accuracy')
net2.fit(X,y,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 1.3192 - accuracy: 0.7984
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.4260 - accuracy: 0.8600
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3790 - accuracy: 0.8711
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3572 - accuracy: 0.8771
Epoch 5/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3407 - accuracy: 0.8831


In [48]:
net2.evaluate(XX,yy)

313/313 [==============================] - 1s 3ms/step - loss: 0.4007 - accuracy: 0.8666


[0.4006589353084564, 0.866599977016449]

In [49]:
net2.summary()

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_5 (Conv2D)           (None, 27, 27, 30)        150       
                                                                 
 max_pooling2d_4 (MaxPooling  (None, 13, 13, 30)       0         
 2D)                                                             
                                                                 
 conv2d_6 (Conv2D)           (None, 12, 12, 30)        3630      
                                                                 
 max_pooling2d_5 (MaxPooling  (None, 6, 6, 30)         0         
 2D)                                                             
                                                                 
 flatten_9 (Flatten)         (None, 1080)              0         
                                                                 
 dense_23 (Dense)            (None, 10)              

dense에 파라메터 뭉쳐있는 이유
- 펼쳐진 노드를 뭉치기 때문에..
- 1080*10*bias10개 = 10810 이렇게 계산되지!

- 정확도만 보면 net2가 낫네?
- 파라메터 수를 보니 첫번째 시도 대비 파라메터를 적게 사용한 모습을 볼 수 있다.

$\therefore$ 파라메터 조금쓰고 정확도 높은 좋은 net2모형!

In [50]:
net2.layers

내가 설계한 네트워크의 순서를 확인해보자

In [51]:
c1,m1,c2,m2,flttn,dns = net2.layers

In [56]:
print(X.shape) # 입력이미지 = 2D # 1은 채널이 하나라는 뜻
print(c1(X).shape) # 2D # 갑자기 채널이 30이 되어버리!
print(m1(c1(X)).shape) # 2D # maxpool 통과했더니 fixel 크기가 줄어들었네?
print(c2(m1(c1(X))).shape) # 2D # 또 줄어들고 채널은 30으로 유지되는데?
print(m2(c2(m1(c1(X)))).shape) # 2D # 모두 퍼져서 12*12*30=4320이 되어버림
print(flttn(m2(c2(m1(c1(X))))).shape) # 2D
print(dns(flttn(m2(c2(m1(c1(X)))))).shape) # 2D

(60000, 28, 28, 1)
(60000, 27, 27, 30)
(60000, 13, 13, 30)
(60000, 12, 12, 30)
(60000, 6, 6, 30)
(60000, 1080)
(60000, 10)


In [41]:
net2 = tf.keras.Sequential()
net2.add(tf.keras.layers.Conv2D(30,(2,2),activation='relu'))
net2.add(tf.keras.layers.MaxPool2D())
net2.add(tf.keras.layers.Conv2D(30,(2,2),activation='relu'))
net2.add(tf.keras.layers.MaxPool2D())
net2.add(tf.keras.layers.Flatten())
net2.add(tf.keras.layers.Dense(10,activation='softmax'))
net2.compile(optimizer='adam',loss=tf.losses.categorical_crossentropy,metrics='accuracy')
net2.fit(X,y,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.7736 - accuracy: 0.8055
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3742 - accuracy: 0.8654
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3343 - accuracy: 0.8789
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.3126 - accuracy: 0.8867
Epoch 5/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2971 - accuracy: 0.8908


In [42]:
net2.evaluate(XX,yy)

313/313 [==============================] - 1s 3ms/step - loss: 0.3276 - accuracy: 0.8802


[0.32759028673171997, 0.8802000284194946]

## maxpoling

#### 테스트1

In [59]:
m = tf.keras.layers.MaxPooling2D()

In [62]:
XXX = tnp.arange(1*4*4*1).reshape(1,4,4,1)
XXX # observation / color / channel

<tf.Tensor: shape=(1, 4, 4, 1), dtype=int64, numpy=
array([[[[ 0],
         [ 1],
         [ 2],
         [ 3]],

        [[ 4],
         [ 5],
         [ 6],
         [ 7]],

        [[ 8],
         [ 9],
         [10],
         [11]],

        [[12],
         [13],
         [14],
         [15]]]])>

In [64]:
m(XXX).reshape(1,2,2)

<tf.Tensor: shape=(1, 2, 2), dtype=int64, numpy=
array([[[ 5,  7],
        [13, 15]]])>

여기로 가는 mapping을 찾아볼까

In [63]:
XXX.reshape(1,4,4) #우리가 생각하는 이미지의 모습

<tf.Tensor: shape=(1, 4, 4), dtype=int64, numpy=
array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]]])>

[0,1,4,5] 중 max는 5

[2,3,6,7] 중 max는 7

[8,9,12,13] 중 max는 13

[10,11,14,15] 중 max는 15

`-` MaxPool2D layer의 역할: (2,2) 윈도우를 만들고 (2,2) 윈도우에서 max를 뽑아 값을 기록, 윈도우를 움직이면서 반복

#### 테스트2

In [65]:
XXX = tnp.arange(1*6*6*1).reshape(1,6,6,1)
XXX.reshape(1,6,6) # 흑백 이미지라면 이렇게 보일 것

<tf.Tensor: shape=(1, 6, 6), dtype=int64, numpy=
array([[[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23],
        [24, 25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34, 35]]])>

In [66]:
m(XXX).reshape(1,3,3)

<tf.Tensor: shape=(1, 3, 3), dtype=int64, numpy=
array([[[ 7,  9, 11],
        [19, 21, 23],
        [31, 33, 35]]])>

#### 테스트3

In [71]:
#collapse_output
tf.keras.layers.MaxPooling2D??

Init signature:
tf.keras.layers.MaxPooling2D(
    pool_size=(2, 2),
    strides=None,
    padding='valid',
    data_format=None,
    **kwargs,
)
Source:        
@keras_export('keras.layers.MaxPool2D', 'keras.layers.MaxPooling2D')
class MaxPooling2D(Pooling2D):
  """Max pooling operation for 2D spatial data.

  Downsamples the input along its spatial dimensions (height and width)
  by taking the maximum value over an input window
  (of size defined by `pool_size`) for each channel of the input.
  The window is shifted by `strides` along each dimension.

  The resulting output,
  when using the `"valid"` padding option, has a spatial shape
  (number of rows or columns) of:
  `output_shape = math.floor((input_shape - pool_size) / strides) + 1`
  (when `input_shape >= pool_size`)

  The resulting output shape when using the `"same"` padding option is:
  `output_shape = math.floor((input_shape - 1) / strides) + 1`

  For example, for `strides=(1, 1)` and `padding="valid"`:

  >>> x = tf.con

In [72]:
m = tf.keras.layers.MaxPooling2D(pool_size=(3,3))

In [73]:
XXX = tnp.arange(1*6*6*1).reshape(1,6,6,1)
XXX.reshape(1,6,6) # 흑백 이미지라면 이렇게 보일 것

<tf.Tensor: shape=(1, 6, 6), dtype=int64, numpy=
array([[[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23],
        [24, 25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34, 35]]])>

In [74]:
m(XXX).reshape(1,2,2)

<tf.Tensor: shape=(1, 2, 2), dtype=int64, numpy=
array([[[14, 17],
        [32, 35]]])>

#### 테스트4

In [77]:
m = tf.keras.layers.MaxPooling2D(pool_size=(2,2))

In [78]:
XXX = tnp.arange(1*5*5*1).reshape(1,5,5,1)
XXX.reshape(1,5,5) # 흑백 이미지라면 이렇게 보일 것

<tf.Tensor: shape=(1, 5, 5), dtype=int64, numpy=
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24]]])>

In [79]:
m(XXX).reshape(1,2,2)

<tf.Tensor: shape=(1, 2, 2), dtype=int64, numpy=
array([[[ 6,  8],
        [16, 18]]])>

애매한 열/행 수를 넣어주니 마지막 열/행은 버려버리네!
- 차원이 안 맞으면 버려버림( 기본 옵션)

In [83]:
m = tf.keras.layers.MaxPooling2D(pool_size=(2,2),padding='same')

In [84]:
XXX = tnp.arange(1*5*5*1).reshape(1,5,5,1)
XXX.reshape(1,5,5) # 흑백 이미지라면 이렇게 보일 것

<tf.Tensor: shape=(1, 5, 5), dtype=int64, numpy=
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24]]])>

In [86]:
m(XXX).reshape(1,3,3)

<tf.Tensor: shape=(1, 3, 3), dtype=int64, numpy=
array([[[ 6,  8,  9],
        [16, 18, 19],
        [21, 23, 24]]])>

`padding ="same"` 옵션을 넣어주니 차원을 버리지 않고 다 사용하네?
- 균형을 맞춰 구하진 않음
- 24의 경우 24 하나의 값 중에서 큰 값인 24를 택한 것.

#### 테스트5

In [87]:
XXX = tnp.arange(2*4*4*1).reshape(2,4,4,1)
XXX.reshape(2,4,4)

<tf.Tensor: shape=(2, 4, 4), dtype=int64, numpy=
array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]],

       [[16, 17, 18, 19],
        [20, 21, 22, 23],
        [24, 25, 26, 27],
        [28, 29, 30, 31]]])>

In [88]:
m(XXX).reshape(2,2,2)

<tf.Tensor: shape=(2, 2, 2), dtype=int64, numpy=
array([[[ 5,  7],
        [13, 15]],

       [[21, 23],
        [29, 31]]])>

#### 테스트5

In [89]:
XXX = tnp.arange(1*4*4*3).reshape(1,4,4,3) # observation 1 channel 3으로 고정

In [92]:
XXX[:,:,:,0]

<tf.Tensor: shape=(1, 4, 4), dtype=int64, numpy=
array([[[ 0,  3,  6,  9],
        [12, 15, 18, 21],
        [24, 27, 30, 33],
        [36, 39, 42, 45]]])>

위 아래 같은 결과..! 생략구문이라 부른다.

In [93]:
XXX[...,0]

<tf.Tensor: shape=(1, 4, 4), dtype=int64, numpy=
array([[[ 0,  3,  6,  9],
        [12, 15, 18, 21],
        [24, 27, 30, 33],
        [36, 39, 42, 45]]])>

In [91]:
m(XXX)[:,:,:,0]

<tf.Tensor: shape=(1, 2, 2), dtype=int64, numpy=
array([[[15, 21],
        [39, 45]]])>

In [94]:
m(XXX)[...,0]

<tf.Tensor: shape=(1, 2, 2), dtype=int64, numpy=
array([[[15, 21],
        [39, 45]]])>